pages/api/post/[id].ts

```diff
@@ -0,0 +1,206 @@
+  req: NextApiRequest,
+  res: NextApiResponse,
+  next: () => void
+) {
+  const postId: string | null = getIdFromRequest(req, res);
+  if (postId) {
+    (req as NextApiRequestWithPostId).postId = postId;
+  } else {
+    res.status(StatusCodes.BAD_REQUEST).end();
+  }
+  next();
+}
+
+async function getHandler(
+  req: NextApiRequestWithPostId,
+  res: NextApiResponse
+): Promise<void> {
+  if (!req.postId) {
+    return;
+  }
+  const includeUser: boolean = await isAuthenticated(req, res);
+  const post: Post | PostWithUser | null = await getPost(
+    req.postId,
+    includeUser
+  );
+  if (post) {
+    res //
+      .status(StatusCodes.OK) //
+      .json(await getPost(req.postId, includeUser));
+    res.end();
+    return;
+  }
+  res.status(StatusCodes.NOT_FOUND).end();
+}
+
+async function putHandler(
+  req: NextApiRequestWithPostIdAndImage,
+  res: NextApiResponse
+): Promise<void> {
+  if (!req.postId) {
+    return;
+  }
+  if (!(await isPostFromUser(req.postId, req, res))) {
+    return;
+  }
+
+  const updatablePost: UpdatablePost | null = getUpdatablePostFromRequest(req);
+  if (!updatablePost) {
+    res.status(StatusCodes.BAD_REQUEST).end();
+    return;
+  }
+
+  const updatedPost: Post = await updatePost(req.postId, updatablePost);
+  res //
+    .status(StatusCodes.OK) //
+    .json(updatedPost);
+}
+
+async function deleteHandler(
+  req: NextApiRequestWithPostId,
+  res: NextApiResponse
+): Promise<void> {
+  if (!req.postId) {
+    return;
+  }
+  if (!(await isPostFromUser(req.postId, req, res))) {
+    return;
+  }
+  await deletePost(req.postId);
+  res.status(StatusCodes.OK).end();
+}
+
+function getIdFromRequest(
+  req: NextApiRequest,
+  res: NextApiResponse
+): string | null {
+  const { id } = req.query;
+  if (Array.isArray(id)) {
+    res.status(StatusCodes.BAD_REQUEST).end();
+    return null;
+  }
+  const postId = id as string;
+  if (!validate(postId)) {
+    res.status(StatusCodes.BAD_REQUEST).end();
+    return null;
+  }
+  return postId;
+}
+
+async function isPostFromUser(
+  postId: string,
+  req: NextApiRequest,
+  res: NextApiResponse
+): Promise<boolean> {
+  const postWithUser: PostWithUser | null = (await getPost(
+    postId,
+    true
+  )) as PostWithUser | null;
+  if (!postWithUser) {
+    res.status(StatusCodes.NOT_FOUND).end();
+    return false;
+  }
+
+  const user: User | null = await getCurrentUser(req, res);
+  if (!user) {
+    res.status(StatusCodes.UNAUTHORIZED).end();
+    return false;
+  } else if (user.id !== postWithUser.userId) {
+    res.status(StatusCodes.FORBIDDEN).end();
+    return false;
+  }
+
+  return true;
+}
+
+function getUpdatablePostFromRequest(
+  req: NextApiRequestWithPostIdAndImage
+): UpdatablePost | null {
+  if (!req.body) {
+    return null;
+  }
+
+  const { price, description, status } = req.body;
+  const image = req.file;
+
+  if (Array.isArray(description)) {
+    return null;
+  }
+
+  if (Array.isArray(status)) {
+    return null;
+  }
```

Comment:
What if the someone supplies an array anyways? We need to return a 400 then, right? I think I don't understand something here.

Reply:
I think it's better to check that the supplied body is the type/structure we want, rather than checking if it's a type we don't want, like this:

https://github.com/calebmech/Book-Bazar/blob/07d366c80b372d1fadbcf9af79a641b191fdac98/pages/api/auth/magic.ts#L21-L30

That way we don't have to check through an exhaustive list of types

Reply:
Okay. That makes sense to me. I'll do this.